packages/core/src/compiler/Style.ts

```diff
@@ -2633,6 +2853,105 @@ export const disambiguateFunctions = (env: Env, subProg: SubProg) => {
   subProg.statements.forEach((stmt: SubStmt) => disambiguateSubNode(env, stmt));
 };
 
+//#region Checking translation
+
+const isStyErr = (res: TagExpr<VarAD> | IFGPI<VarAD> | StyleError): boolean =>
+  res.tag !== "FGPI" && !isTagExpr(res);
+
+const findPathsExpr = (expr: Expr): Path[] => {
+  // TODO: Factor the expression-folding pattern out from here and `checkBlockExpr`
+  if (isPath(expr)) {
+    return [expr];
+  } else if (
+    expr.tag === "CompApp" ||
+    expr.tag === "ObjFn" ||
+    expr.tag === "ConstrFn"
+  ) {
+    return _.flatMap(expr.args, findPathsExpr);
+  } else if (expr.tag === "BinOp") {
+    return _.flatMap([expr.left, expr.right], findPathsExpr);
+  } else if (expr.tag === "UOp") {
+    return findPathsExpr(expr.arg);
+  } else if (
+    expr.tag === "List" ||
+    expr.tag === "Vector" ||
+    expr.tag === "Matrix"
+  ) {
+    return _.flatMap(expr.contents, findPathsExpr);
+  } else if (expr.tag === "ListAccess") {
+    return [expr.contents[0]];
+  } else if (expr.tag === "GPIDecl") {
+    return _.flatMap(
+      expr.properties.map((p) => p.value),
+      findPathsExpr
+    );
+  } else if (expr.tag === "Layering") {
+    return [expr.below, expr.above];
+  } else if (expr.tag === "PluginAccess") {
+    return _.flatMap([expr.contents[1], expr.contents[2]], findPathsExpr);
+  } else if (expr.tag === "Tuple") {
+    return _.flatMap([expr.contents[0], expr.contents[1]], findPathsExpr);
+  } else if (expr.tag === "VectorAccess") {
+    return [expr.contents[0]].concat(findPathsExpr(expr.contents[1]));
+  } else if (expr.tag === "MatrixAccess") {
+    return [expr.contents[0]].concat(
+      _.flatMap(expr.contents[1], findPathsExpr)
+    );
+  } else if (
+    expr.tag === "Fix" ||
+    expr.tag === "Vary" ||
+    expr.tag === "StringLit" ||
+    expr.tag === "BoolLit"
+  ) {
+    return [];
+  } else {
+    console.error("expr", expr);
+    throw Error("unknown tag");
+  }
+};
+
+// Find all paths given explicitly anywhere in an expression in the translation.
+// (e.g. `x.shape above y.shape` <-- return [`x.shape`, `y.shape`])
+const findPathsField = (
+  name: string,
+  field: Field,
+  fexpr: FieldExpr<VarAD>,
+  acc: Path[]
+): Path[] => {
+  if (fexpr.tag === "FExpr") {
+    // Only look deeper in expressions, because that's where paths might be
+    if (fexpr.contents.tag === "OptEval") {
+      const res: Path[] = findPathsExpr(fexpr.contents.contents);
+      return acc.concat(res);
+    } else {
+      return acc;
+    }
+  } else if (fexpr.tag === "FGPI") {
+    // Get any exprs that the properties are set to
+    const propExprs: Expr[] = Object.entries(fexpr.contents[1])
+      .map((e) => e[1])
+      .filter((e: TagExpr<VarAD>): boolean => e.tag === "OptEval")
```

Comment:
You can do this in one single lambda function, so TS would know the type of `e` is changed to `OptEval` (in an `if` statement, for instance).

Reply:
Not sure how I would do this in a lambda in this context, as I would need to be doing both a map and a filter in it. But can do this in the future.

Reply:
Just saw the `filter`. Don't really know a great alternative :(.